### ISSUE TITLE
create readonly and constant variable easily

### ISSUE THREAD

**Issue by p0W3RH311** [PowerShell/PowerShell#15044]
Hi Guys !!

we can create readonly and constant variable with new-variable and set-variable cmdlets but its much easily and _natural_ to use this syntax :
```

[ReadOnly]$var1 = 1
[Constant]$var2 = 2

```

**Comment by powercode** [comment]
That is the syntax used for types. I think it would be confusing and ambiguous to use that.
If you have a type named ReadOnly or Constant, which is completely legal, you can't tell that apart from variable properties.

**Comment by p0W3RH311** [comment]
[PSReadOnly]
[PSConstant]

**Comment by mklement0** [comment]
I think @powercode is correct: we should reserve the `[<name>]` syntax for type literals to avoid confusion and name conflicts (we already have one unfortunate exception, pseudo-type `[ordered]`; I don't think we should add more.)

I do think it would be nice to have a simpler way to declare such variables, however.

Given that `$private:foo` already declares an `-Option Private` variable, one option is to support `$readonly:foo` and `$constant:foo`.

Due to support for namespace variable notation, there too the potential for naming conflict exists, if someone happens to declare PS drives named `readonly` and `constant` (which doesn't seem likely, however).

**Comment by jborean93** [comment]
We can definitely have attributes on a var like `[ValidateNotNull()]$var` so extending that further like `[ReadOnly()]` or `[Constant()]` doesn't seem too hard of an ask.

**Comment by vexx32** [comment]
For me at least, a readonly/constant attribute seems to make more sense than a pseudo-namespace.

That said, the utility of this kind of thing in PowerShell is generally fairly limited, so I'm not sure it's really worth the necessary time investment to implement.

**Comment by mklement0** [comment]
> more sense than a pseudo-namespace.

My thinking is twofold:

* We already have such pseudo namespaces / scope specifiers: `$private:`, `$using:`, `$global:`, `$script:`, `$local:` - they wont go away, and `$readonly:` and `$constant:` fit right in.
   * `$private:` isn't a _scope_. it is a variable _option_ applied to a _local_ variable, as would `$readonly:` and `$constant:` be; this conflation isn't great, but it is well-established.

* By contrast, even though the use of attributes ultimately provides more flexibility:
  * Using attributes is an advanced technique rarely seen in the wild, probably because people don't _expect_ it to work.
  * The syntax doesn't come naturally to those used to PowerShell-native features only.

**Comment by vexx32** [comment]
I don't agree. We currently have scope modifiers. `$private:` is the only real outlier, but that's mainly an implementation detail. All the current scope modifiers are just that.

"readonly" is not a scope. Nor is "constant". Neither of those properties are scopes. They're properties of a variable.

And yeah, it won't come naturally. _**Good**_. There are irritating caveats and weirdness associated with using readonly or constant variables. Folks who use them should be well aware before they stumble into a day's worth of needless confusion resulting from careless use.

**Comment by mklement0** [comment]
> There are irritating caveats and weirdness associated with using readonly or constant variables. 

That's a separate discussion. 
We're not debating the inherent merits of PowerShell's readonly / constant variables here - only how to facilitate _declaring_ them.

The general point is: `$<name>:` is already a muddled conflation of several concepts in that, `<name>`:

* can refer to a _drive_, e.g., `$function:`

* can refer to a bona fide _scope_: `$global:`, `$local:`, `$script:`

* can refer to a _quasi_-scope - a reference to the caller's scope _tree_ (something either _directly_ in the caller's scope or _visible_ in the caller's scope via an ancestral scope), via `$using:`

* yes,  currently an outlier, a local-scope variable with _option_ `Private`, preventing it from seen by descendant scopes (`$scope:`)

Yes, I can see how "we've-already-created-this-mess-so-let's-add-to-it"  isn't a great argument in general.

Pragmatically speaking, however, adding `$readonly:` and `$constant:` works in a kinda-sorta-if-you-squint-just-right way, and to my mind better than venturing into the world of attributes.

And, to recap, arguing that this shouldn't be made easy because you don't like the underlying concepts is (a)  specious in itself and (b) debating these concepts is an entirely separate conversation.

To put it differently: if you think the underlying concepts aren't worthwhile, we shouldn't be discussing the best syntax to implement them.

**Comment by mklement0** [comment]
Let me try to frame it another way:

* If we think that read-only and constant variables are too treacherous a feature to spring on unsuspecting users:
  * We should debate these features and see if they're redeemable.
  * Before a conclusion is reached, any discussion about _syntactic sugar_ should be _shelved_.

* Once a conclusion has been reached:
  * If the features are redeemable, let's discuss syntactic sugar: attributes vs. pseudo scopes.
  * If not, the existing `New-Variable` and `Set-Variable` cmdlets will do.

**Comment by mklement0** [comment]
To return to the syntactic sugar debate:

> "readonly" is not a scope. Nor is "constant". Neither of those properties are scopes. They're properties of a variable.

In PowerShell terms, they are variable _options_ ([`System.Management.Automation.ScopedItemOptions`](https://docs.microsoft.com/en-US/dotnet/api/System.Management.Automation.ScopedItemOptions)) - just like `Private` is, _for which we already have a "scope" specifier_, `$private:` That said, unlike `ReadOnly` and `Constant`, `Private` has scope-behavior _implications_, as does `AllScope`.

**Comment by SeeminglyScience** [comment]
Related/half dupe of #8906

Also I'm not sure it makes sense to put any effort into this until they're reworked.  The way they were designed and implemented doesn't really make sense outside of automatic variables, and makes accidental shadowing even more of an issue.

**Comment by mklement0** [comment]
Thanks for that link, @SeeminglyScience - hadn't seen it; interesting debate.

Since the linked issue is open and has actually been green-lighted - based on the attributes approach - but doesn't mention shadowing or other pitfalls, perhaps you can flesh these out there - or perhaps create a new issue, if you have specific ideas for how to improve how these variables work. (I get that accidental variable shadowing is a general concern in PowerShell).

**Comment by SeeminglyScience** [comment]
> Since the linked issue is open and has actually been green-lighted - based on the attributes approach

FWIW my understanding was that they asserted what the right pattern would be, but have not ruled on if it would be accepted.  Maybe that's implied though, I dunno.

> perhaps you can flesh these out there - or perhaps create a new issue, if you have specific ideas for how to improve how these variables work. (I get that accidental variable shadowing is a general concern in PowerShell).

I went over it a little bit in [this comment](https://github.com/PowerShell/PowerShell/issues/8906#issuecomment-531911209).  Basically I'd like to see these as parse time concepts.

For constants, the compiler would actually treat them as a constant value rather than a variable lookup. This would greatly limit what types of variables or expressions could be used (similar to C#), but it could provide some solid readability (variables in attributes for instance) and somewhat of a performance boost.

For readonly, I'd like to see something that does *not* set the `ScopedItemOptions`, but instead just tells the parser to yell if it sees a reassignment. That solves the shadowing problem, and gives immediate feedback in the editor when accidently shadowing in the same lexical scope.

Would these changes be breaking? Yeah. Are they worth doing anyway? Ehhh... probably not.  I don't think it'd get much use still and it'd be a lot of work.

**Comment by mklement0** [comment]
> but have not ruled on if it would be accepted.
> Maybe that's implied though, I dunno.

Yes, I believe the label "Up-for-Grabs" implies that.
@iSazonov, can you confirm?

The improvements you suggest make sense, and I can see the potential problem with the effort-reward balance (as well as the potential to break things). The bigger, fundamental problem, I think, is PowerShell's dynamic scoping and shadow-on-assignment logic - but changing that would be the mother of all breaking changes.

A potential middle ground that requires much less effort is to make the attribute turn on _both_ `AllScope` and `ReadOnly` / `Constant` - which is supported, but not directly via the syntax of `New-Variable` and `Set-Variable`, because the flag nature of the `ScopedItemOptions` enum isn't supported due to only supporting a _scalar_ `-Option` argument.

To demonstrate:

```powershell
# Note: You can't do this with *just* New-Variable / Set-Variable
New-Variable foo bar -Option AllScope; (Get-Variable foo).Options += 'ReadOnly'

& { $foo = 'baz' } # FAILS even in CHILD SCOPE, as intended: "WriteError: Cannot overwrite variable foo because it is read-only or constant."
```

This would solve the shadowing problem.

Would PSES be able to pick up on this?

Going back to the syntax and potential implementation, and reviving @p0W3RH311's idea of using _types_ rather than _attributes_, which do strike me as syntactically preferable:

`[PSReadOnly]` and `[PSConstant]` could then be new type accelerators that refer to two new classes that could derive from [`System.Management.Automation.PSVariable`](https://docs.microsoft.com/en-US/dotnet/api/System.Management.Automation.PSVariable), `System.Management.Automation.PSReadOnlyVariable`  and `System.Management.Automation.PSConstantVariable`  (somewhat paradoxically), which could encapsulate the `ReadOnly | AllScope` and `Constant | AllScope` construction respectively, which would give us:

```powershell
[PSReadOnly] $foo = 'bar'
& { $foo = 'baz' } # FAILS
```

**Comment by mklement0** [comment]
P.S.: Of course, this logic could also be hidden behind `$readonly:` and `$constant:`.

To me, the only thing the attribute-based solution has going for it is that it bears the _lowest_ risk of name conflicts - for the developer experience, I think `[PSReadOnly]` and `$readonly:` are preferable.

Either way, we should fix `New-Variable` and `Set-Variable` to allow direct construction of `ReadOnly | AllScope` and `Constant | AllScope` variables - in the simplest case by changing `-Option <ScopedItemOptions>` to `-Option <ScopedItemOptions[]>`.

**Comment by SeeminglyScience** [comment]
> This would solve the shadowing problem.

Or cause it, depending on how you look at it. There's two sides to the shadowing problem

1. Variable shadowing within the same project/script. `AllScope` deals with that somewhat
2. Shadowing between different non-module projects/scripts. If you call a build script that happens to use the same variable, it now blows up

That's why I think it should be a semantic check in the parser rather than something the runtime enforces.

> Would PSES be able to pick up on this?

PSSA could hack together an analyzer, but it'd be very fragile.

**Comment by mklement0** [comment]
> Shadowing between different non-module projects/scripts.

Dynamic scoping: the gift that keeps on giving - and which we cannot ungift without serious breaking changes.

>  a semantic check in the parser 

If I understand correctly, this wouldn't _solve_ that problem, it would only report it _earlier_ (parsing script B in which `$foo` is assigned to after having executed script A that defined a read-only `$foo`)  - and that to me indeed doesn't seem worth the effort.

A proper solution will have to wait till #6745, I think.

In the meantime, the `ReadOnly | AllScope` solution provides at least _some_ value and is comparatively easy to implement and, as an _opt-in_, is non-breaking - you just need to be aware of the pitfall you describe.

**Comment by SeeminglyScience** [comment]
> Dynamic scoping: the gift that keeps on giving - and which we cannot ungift without serious breaking changes.

😁

> If I understand correctly, this wouldn't _solve_ that problem, it would only report it _earlier_ (parsing script B in which `$foo` is assigned to after having executed script A that defined a read-only `$foo`; it also wouldn't help at _edit_ time) - and that to me indeed doesn't seem worth the effort.

The semantic check would be purely static analysis.

So for example:

```powershell
$sb = {
    $testing = 'something'
}

function Foo {
    param([scriptblock] $Action)
    end {
        [readonly()] $testing = ''

        # Doesn't throw here, code is external
        & $Action

        # Parser error here.  It can determine that it's a reassignment of a readonly variable.
        # No runtime checks needed.
        $testing = 'a'
}

Foo $sb
```

Think of it like how in a class, if you try to use a variable that wasn't explicitly assigned in the method, you get a parse error.  Same type of thing, runtime state doesn't matter, it's purely analysis.

> In the meantime, the `ReadOnly | AllScope` solution provides at least _some_ value and is comparatively easy to implement and, as an _opt-in_, is non-breaking - you just need to be aware of the pitfall you describe.

It solves the problem that is the least concerning imo.  I still very strongly recommend against this pattern for anything other than automatic variables.

**Comment by mklement0** [comment]
Quick follow-up on the `-Option` typing: I forgot that even a scalar `ScopedItemOptions` allows you to pass _multiple_ flags, courtesy of PowerShell's flexible from-string conversion: `New-Variable foo bar -Option 'ReadOnly, AllScope'` works; however, because it is obscure, it still makes sense to make the parameter array-valued (for the more natural `New-Variable foo bar -Option ReadOnly, AllScope` syntax) - see #15057

**Comment by mklement0** [comment]
As a quick aside, @SeeminglyScience: Thanks for showing me that multiple values can actually be passed as an _array_ to a flag-based enum parameter type, even if the parameter type is a scalar; that is,
`New-Variable foo bar -Option ReadOnly, AllScope` works just fine - I've created an issue to suggest better user guidance in syntax diagrams, parameter names, and the docs in that regard: #15058

**Comment by mklement0** [comment]
Understood re static analysis; I made an assumptions that you didn't make, namely that the variable is declared with `AllScope`, in which case the per-script static analysis doesn't help in the script-calling-another-script scenario (and if the static analysis of script B doesn't take into account the state created by calling script A at the time of invocation of B, then even the earlier-detection bonus doesn't apply).

With per-script-file-in-isolation static analysis all you get is early warning _per script_, and _cross-script_  you're again faced with:

* potential inadvertent shadowing if `AllScope` is _not_ used.
* potential unwanted breakage if `AllScope` _is_ used.

Such static analysis would amount to a limited "lexical scoping overlay" over PowerShell's dynamic scoping that to me doesn't seem worth the - nontrivial, I'm sure - effort.
I think a hypothetical future PowerShell version with lexical scoping only would be the only satisfactory solution.

---

In terms of providing guidance to script authors who want to make do with the _current_ features, it boils down to this:

* If you want to limit your readonly/constant variable to the _current scope only_, use
`New-Variable foo bar -Option Private, Readonly` - but you need to be aware of what does and doesn't directly in the current scope.

* Otherwise, dynamic scoping applies, and you have two options:

  * If you want your descendant scopes to be able to shadow (override with local copies) your readonly/constant variables, use
`New-Variable foo bar -Option ReadOnly` - but note that _accidental_ shadowing may occur.

  * If you want to make your descendant scopes to use your definition of your readonly/constant variable (though in the readonly case descendant scope can still  forcefully remove and thereby override it), use
`New-Variable foo bar -Option ReadOnly, AllScope` - but note that unrelated code may accidentally try to assign to your variable and therefore malfunction.

This leaves the question whether it's worth throwing syntactic sugar at these use cases.
I can see both sides of the argument, and ultimately don't feel strongly about this.

**Comment by daxian-dbw** [comment]
Engine working group:
This is a duplicate of #8906. The powershell committee has reviewed it and agreed that "attributes is the right solution", so let's go with what we have decided.
